src/OS/QMidi_ALSA.cpp

```diff
@@ -118,4 +126,233 @@ void QMidiOut::sendMsg(qint32 msg)
+    snd_seq_port_info_t* pinfo;
+    int client;
+    int err;
+    snd_seq_t* handle;
+
+    err = snd_seq_open(&handle, "hw", SND_SEQ_OPEN_INPUT, 0);
+    if (err < 0) {
+        /* Use snd_strerror(errno) to get the error here. */
+        return ret;
+    }
+
+    snd_seq_client_info_alloca(&cinfo);
+    snd_seq_client_info_set_client(cinfo, -1);
+
+    while (snd_seq_query_next_client(handle, cinfo) >= 0) {
+        client = snd_seq_client_info_get_client(cinfo);
+        snd_seq_port_info_alloca(&pinfo);
+        snd_seq_port_info_set_client(pinfo, client);
+
+        snd_seq_port_info_set_port(pinfo, -1);
+        while (snd_seq_query_next_port(handle, pinfo) >= 0) {
+            int cap = (SND_SEQ_PORT_CAP_SUBS_READ | SND_SEQ_PORT_CAP_READ);
+            if ((snd_seq_port_info_get_capability(pinfo) & cap) == cap) {
+                QString port = QString::number(snd_seq_port_info_get_client(pinfo));
+                port += ":" + QString::number(snd_seq_port_info_get_port(pinfo));
+                QString name = snd_seq_client_info_get_name(cinfo);
+                ret.insert(port, name);
+            }
+        }
+    }
+
+    snd_seq_close(handle);
+
+    return ret;
+}
+
+bool QMidiIn::connect(QString inDeviceId)
+{
+    if (fConnected)
+    {
+        disconnect();
+    }
+
+    fMidiPtrs = new NativeMidiInInstances;
+    int err = snd_seq_open(&fMidiPtrs->midiIn, "default", SND_SEQ_OPEN_INPUT, 0);
+    if (err < 0)
+    {
+        delete fMidiPtrs;
+        return false;
+    }
+    snd_seq_set_client_name(fMidiPtrs->midiIn, "QMidi");
+    snd_seq_create_simple_port(fMidiPtrs->midiIn, "Input Port", SND_SEQ_PORT_CAP_WRITE,
+                               SND_SEQ_PORT_TYPE_MIDI_GENERIC);
+
+    // connect the device to our previously created port
+    QStringList l = inDeviceId.split(":");
+    int client = l.at(0).toInt();
+    int port = l.at(1).toInt();
+    snd_seq_connect_from(fMidiPtrs->midiIn, 0, client, port);
+
+    fDeviceId = inDeviceId;
+    fConnected = true;
+    return true;
+}
+
+void QMidiIn::disconnect()
+{
+    if (!fConnected)
+    {
+        return;
+    }
+
+    QStringList l = fDeviceId.split(":");
+    int client = l.at(0).toInt();
+    int port = l.at(1).toInt();
+
+    snd_seq_disconnect_to(fMidiPtrs->midiIn, 0, client, port);
+    fConnected = false;
+
+    snd_seq_close(fMidiPtrs->midiIn);
+    delete fMidiPtrs;
+    fMidiPtrs = nullptr;
+}
+
+void QMidiIn::start()
+{
+    if (!fConnected)
+    {
+        return;
+    }
+
+    fMidiPtrs->receiveThread = new MidiInReceiveThread(this, fMidiPtrs);
+    fMidiPtrs->receiveThread->start();
+}
+
+void QMidiIn::stop()
+{
+    if (!fConnected)
+    {
+        return;
+    }
+
+    fMidiPtrs->receiveThread->requestInterruption();
+    fMidiPtrs->receiveThread->wait();
+    fMidiPtrs->receiveThread->deleteLater();
+    fMidiPtrs->receiveThread = nullptr;
+}
+
+MidiInReceiveThread::MidiInReceiveThread(QMidiIn* qMidiIn, NativeMidiInInstances* fMidiPtrs, QObject* parent)
```

Comment:
Since this class is private, it should probably be namespaced or otherwise hidden in a way that it will not conflict with other class names.

Reply:
Guess this should be applied to the `MidiInConsumer` class in the Haiku part as well then?

Reply:
Yes, indeed.